### PR TITLE
feat: Configure proxy container for graceful termination

### LINKED ISSUE #361: Feature Request: always configure pre-stop hook to call /quitquitquit
The proxy exits with exiting with exit code 143 when it receives a SIGTERM from kubernetes.  [cloud-sql-proxy #1803](https://github.com/GoogleCloudPlatform/cloud-sql-proxy/issues/1803) 

To ensure that proxy containers exit with code 0, even when they are 
- Always configure the `--exit-zero-on-sigterm` flag so that the proxy process plays nice with k8s. When  another 
  container on the same pod exits, k8s sends the proxy a SIGTERM, and then the proxy exits with code 0.

- This should be applied to all proxy containers, since we don't want the exit code of the proxy to impact the overall exit status of the pod.

### PR DESCRIPTION
**Pull request by hessjcg**
Add the following configuration to the workload pods so that the proxy can gracefully exit
when the main container is done.

Configure the proxy container to exit 0 when it is terminated. Send a SIGTERM to the proxy container. We want 
the proxy to exit with code 0, indicating a clean termination. Without this change the proxy container would 
exit with exit code 140 (meaning terminated), which would cause kubernetes to report the 
pod as "exited in an error state."

Configure a workload lifecycle handler so that kubernetes calls GET /quitquitquit before terminating the proxy container.
This should give the proxy container the chance to exit gracefully before kubernetes sends a SIGTERM to the 
proxy process. 

Always enable the /quitquitquit proxy api. 

Always set the CSQL_QUIT_URLS environment variable to a space-separated list of proxy quitquitquit urls. This way, when
the main workload container is ready to exit, it can   on workload pods. When a job or cronjob's main process 
is done, that container can iterate over

```
echo Starting job

# execute the job process
run_job

# Tell proxy containers to shut down gracefully
for url in $CSQL_QUIT_URLS ; do 
   wget --post-data '' $url
done
```

Fixes #361

### CODE REVIEW

**Comment by hessjcg** [comment]
Labeled `tests: run` to invoke e2e tests on this PR.

**Review by enocom — COMMENTED** [review]
Do we need to add something to the documentation for this?

**Comment by hessjcg** [comment]
E2E tests passed: https://github.com/GoogleCloudPlatform/cloud-sql-proxy-operator/actions/runs/6052015286/job/16424854308?pr=425

File: internal/workload/podspec_updates.go [hunk 1]
```diff
@@ -810,11 +810,29 @@ func (s *updateState) addHealthCheck(p *cloudsqlapi.AuthProxyWorkload, c *corev1
...
+
+	// For graceful exits as a sidecar, the proxy should exit with exit code 0
+	// when it receives a SIGTERM. The operator therefor configures the proxy
+	// container to enable the /quitquitquit endpoint for graceful exits
+	// and sets the --exit-zero-on-sigterm flag.
+	s.addProxyContainerEnvVar(p, "CSQL_PROXY_QUITQUITQUIT", "true")
+	s.addProxyContainerEnvVar(p, "CSQL_PROXY_EXIT_ZERO_ON_SIGTERM", "true")
+
+	// Configure the pre-stop hook for /quitquitquit
+	c.Lifecycle = &corev1.Lifecycle{
```
Comment (enocom) [inline internal/workload/podspec_updates.go:828]: Should we add a test that verifies this behavior?

**Review by enocom — COMMENTED** [review]
Sounds good on #400. Could we update the title? I'm not sure exactly what the scope is as written.

File: internal/workload/podspec_updates_test.go [hunk 2]
```diff
@@ -911,6 +915,91 @@ func TestPodTemplateAnnotations(t *testing.T) {
 
 }
 
+func TestQuitUrlEnvVar(t *testing.T) {
```
Comment (enocom) [inline internal/workload/podspec_updates_test.go:918]: URL.

Watch out for initialisms.

File: internal/workload/podspec_updates_test.go [hunk 3]
```diff
@@ -911,6 +915,91 @@ func TestPodTemplateAnnotations(t *testing.T) {
...
+		simpleAuthProxy("instance1", "project:server:db"),
+		simpleAuthProxy("instance2", "project:server2:db2"),
+		simpleAuthProxy("instance3", "project:server3:db3")}
+
+	csqls[0].ObjectMeta.Generation = 1
+	csqls[1].ObjectMeta.Generation = 2
+	csqls[2].ObjectMeta.Generation = 3
+
+	var wantQuitUrlsEnv = fmt.Sprintf(
+		"http://localhost:%d/quitquitquit http://localhost:%d/quitquitquit http://localhost:%d/quitquitquit", workload.DefaultAdminPort,
```
Comment (enocom) [inline internal/workload/podspec_updates_test.go:940]: nit: shorter line length is easier to read. how about this?

``` golang
var wantQuitUrlsEnv = strings.Join(
  []string{
    fmt.Sprinf("http://localhost:%d/quitquitquit", workload.DefaultAdminPort),
    fmt.Sprinf("http://localhost:%d/quitquitquit", workload.DefaultAdminPort),
    fmt.Sprinf("http://localhost:%d/quitquitquit", workload.DefaultAdminPort),
  },
  " ",
)

File: internal/workload/podspec_updates_test.go [hunk 4]
```diff
@@ -911,6 +915,91 @@ func TestPodTemplateAnnotations(t *testing.T) {
...
+		t.Fatal("can't find env var", err)
+	}
+	if ev.Value != wantQuitUrlsEnv {
+		t.Fatal("got", ev.Value, "want", wantQuitUrlsEnv)
+	}
+}
+
+func TestPreStopHook(t *testing.T) {
+
+	var (
```
Comment (enocom) [inline internal/workload/podspec_updates_test.go:962]: var u = ...

File: internal/workload/podspec_updates_test.go (hunk 2 above)
Comment (hessjcg) [reply internal/workload/podspec_updates_test.go]: Fixed.

File: internal/workload/podspec_updates_test.go (hunk 3 above)
Comment (hessjcg) [reply internal/workload/podspec_updates_test.go]: Fixed.

File: internal/workload/podspec_updates_test.go (hunk 4 above)
Comment (hessjcg) [reply internal/workload/podspec_updates_test.go]: Fixed.

**Review by hessjcg — COMMENTED** [review]
Updated based on Eno's comments.

**Review by enocom — COMMENTED** [review]
LGTM

We need to pin the next release of the proxy before we can merge this, right?

**Comment by hessjcg** [comment]
No, this will work with the current proxy version and represents an improvement on the proxy. 

It will work even better when the proxy is released with [PR# 1947](https://github.com/GoogleCloudPlatform/cloud-sql-proxy/pull/1947)

**Review by enocom — APPROVED** [review]

**Pull request merged**